File: babydata/src/main/java/com/zero/babydata/room/SleepRecordDao.java
package com.zero.babydata.room;

import androidx.room.Dao;
import androidx.room.Insert;
import androidx.room.Delete;
import androidx.room.Query;

import com.zero.babydata.entity.SleepRecord;

import java.util.List;

@Dao
public interface SleepRecordDao {

    @Insert
    void insertSleepRecord(SleepRecord sleepRecord);

    @Delete
    void deleteSleepRecord(SleepRecord sleepRecord);

    @Query("SELECT * FROM Sleep_Records WHERE babyId = :babyId")
    List<SleepRecord> getAllSleepRecords(int babyId);

    @Query("DELETE FROM Sleep_Records WHERE sleepId = :sleepId")
    void deleteSleepRecordById(int sleepId);

    @Query("SELECT * FROM Sleep_Records WHERE babyId = :babyId AND sleepStart BETWEEN :startOfDay AND :endOfDay ORDER BY sleepStart DESC")
    List<SleepRecord> getSleepRecordsForDay(int babyId, long startOfDay, long endOfDay);

    @Query("SELECT * FROM Sleep_Records WHERE babyId = :babyId ORDER BY sleepEnd DESC LIMIT 1")
    SleepRecord getLastSleepRecord(int babyId);

    @Query("SELECT * FROM Sleep_Records WHERE babyId = :babyId ORDER BY sleepStart DESC LIMIT :limit")
    List<SleepRecord> getRecentSleeps(int babyId, int limit);
}

File: babydata/src/main/java/com/zero/babydata/room/SleepRecordDao.kt
package com.zero.babydata.room

import androidx.room.Dao
import androidx.room.Delete
import androidx.room.Insert
import androidx.room.Query
import androidx.room.Update
import com.zero.babydata.entity.SleepRecord

@Dao
interface SleepRecordDao {

    @Insert
    fun insertSleepRecord(sleepRecord: SleepRecord)

    @Update
    fun updateSleepRecord(sleepRecord: SleepRecord)

    @Delete
    fun deleteSleepRecord(sleepRecord: SleepRecord)

    @Query("SELECT * FROM Sleep_Records WHERE babyId = :babyId")
    fun getAllSleepRecords(babyId: Int): List<SleepRecord>

    @Query("DELETE FROM Sleep_Records WHERE sleepId = :sleepId")
    fun deleteSleepRecordById(sleepId: Int)

    @Query("SELECT * FROM Sleep_Records WHERE babyId = :babyId AND sleepStart BETWEEN :startOfDay AND :endOfDay ORDER BY sleepStart DESC")
    fun getSleepRecordsForDay(babyId: Int, startOfDay: Long, endOfDay: Long): List<SleepRecord>

    @Query("SELECT * FROM Sleep_Records WHERE babyId = :babyId ORDER BY sleepEnd DESC LIMIT 1")
    fun getLastSleepRecord(babyId: Int): SleepRecord?

    @Query("SELECT * FROM Sleep_Records WHERE babyId = :babyId ORDER BY sleepStart DESC LIMIT :limit")
    fun getRecentSleeps(babyId: Int, limit: Int): List<SleepRecord>

    @Query("SELECT DISTINCT date(sleepStart / 1000, 'unixepoch', 'localtime') as recordDate FROM Sleep_Records WHERE babyId = :babyId AND sleepStart BETWEEN :startTime AND :endTime")
    fun getDatesWithSleeps(babyId: Int, startTime: Long, endTime: Long): List<String>

    @Query("SELECT * FROM Sleep_Records WHERE sleepId = :sleepId")
    fun getSleepRecordById(sleepId: Int): SleepRecord?
}
